velox/expression/PeeledEncoding.cpp

```diff
@@ -121,6 +121,7 @@ bool PeeledEncoding::peelInternal(
       if (leaf->isLazy() && leaf->asUnchecked<LazyVector>()->isLoaded()) {
         auto lazy = leaf->asUnchecked<LazyVector>();
         leaf = lazy->loadedVectorShared();
+        setPeeled(leaf, fieldIndex, peeledVectors);
```

Comment:
The PeeledEncoding case that deals with lazy is only when the lazy layer is between 'wraps', however with this change we are also unwrapping a lazy layer over flat which means that to be consistent we need to acknowledge lazy as a 'wrap' and return true for both inputs like lazy(flat) or dict(lazy(flat)) with peeledVectors containing (flat). Otherwise, we might end up writing code that assumes this by only considering the dict(lazy(flat)) case and does not handle lazy(flat) style input. 

For eg, The original issue is that Cast does not expect a lazy vector, if we rely on PeeledEncoding to remove this then we always rely on it removing a lazy layer. however, if the input is lazy(flat) then PeeledEncoding returns false (with the current PR), the lazy remains intact and we still hit this issue in Cast.

Moreover, returning 'true' for a lazy(flat) does not make sense because there is no peel to wrap the result with.

I think the right fix here would instead be to handle this case in Cast and call loadedVector() there to get the underlying vector.
What do you think?

Reply:
I see your concern about lazy(flat). I can add the call to load lazy vector after peeling in CastExpr to unblock Presto for now. On the other hand, I saw PeeledEncoding::peel() is used in FieldReference, JsonCastOperator, and Expr.cpp, so all of them would need to manually handle inner-most lazy layer themselves (FieldReference already does it). Ideally, I think it would be better to let PeeledEncoding handle it since VectorEncoding::Simple::LAZY is considered an encoding and user would expect it to be handled too.

Reply:
So I thought about this a bit more, I am curious what expectations we provide when writing UDFs? Although cast  is a special form so it might have different input expectations, we can still consider the actual casting/conversion implementation as a UDF. The reason i bring this up is, if we always ensure that you never get a lazy layer in a vector then we need to build that functionality into the eval loop. Extending Peeling to cover it would mean we are bleeding concepts other than wrappings into its implementation. Moreover, then we can ensure that invariant is held across all expr eval, for UDFs and special forms and cover the additional cases you identified. What do you think?

Reply:
I think the current expectation for UDFs is just input vector can have ANY encoding, except that if a function only has one argument, this argument is guaranteed to be flat or constant. For vector functions, as long as the UDF author decodes the input vector, since DecodedVector loads lazy by default, they should be fine. For simple functions, the SimpleFunctionAdapter decodes the input vector using DecodedVector, so they should be fine too. For functions that have one argument though, the assumption about input being flat or constant may not be true if there is an inner-most lazy layer. If we don't want to let PeeledEncoding handle lazy, we can potentially call loadedVector() before calling VectorFunction::apply().

(JsonCastOperator needs to handle the peeling itself because it is called from CastExpr that is a special form.)